core/controllers/reader.py

```diff
@@ -122,31 +123,67 @@ def classify(exp_id, state, answer, params):
-            best_matched_truth_value = ored_truth_value
-            best_matched_rule_spec = best_rule_spec
-            best_matched_answer_group = answer_group
-            best_matched_answer_group_index = i
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > best_matched_truth_value:
+                    best_matched_truth_value = evaluated_truth_value
+                    best_matched_rule_spec = rule_spec
+                    best_matched_answer_group = answer_group
+                    best_matched_answer_group_index = i
+
+    # Find the maximum soft rule that matches. This is
+    # done by ORing (maximizing) all truth values of all rules over all answer
+    # groups. The group with the highest truth value is considered the best
+    # match.
+    if best_matched_answer_group is None:
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    evaluated_truth_value = rule_domain.evaluate_rule(
+                        rule_spec, input_type, params, normalized_answer, fs)
+                    if evaluated_truth_value > best_matched_truth_value:
+                        best_matched_truth_value = evaluated_truth_value
+                        best_matched_rule_spec = rule_spec
+                        best_matched_answer_group = answer_group
+                        best_matched_answer_group_index = i
+
+    # Run the classifier
+    if best_matched_answer_group is None:
+        sc = classifier_services.StringClassifier()
+        training_examples = [[doc, []] for doc in
+            state.interaction.confirmed_unclassified_answers]
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    training_examples.extend([[doc, [str(i)]] for doc in
+                        rule_spec.inputs['training_data']])
+        if len(training_examples) > 0:
+            sc.load_examples(training_examples)
+            doc_ids = sc.add_examples_for_predicting([answer])
+            predicted_label = sc.predict_label_for_doc(doc_ids[0])
+            if predicted_label != '_default':
+                predicted_answer_group_index = int(predicted_label)
+                predicted_answer_group = state.interaction.answer_groups[
+                    predicted_answer_group_index]
+                best_matched_truth_value = rule_domain.CERTAIN_TRUE_VALUE
+                for rule_spec in answer_group.rule_specs:
+                    if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
```

Comment:
This part is a bit concerning (the idea of there only being exactly zero or one fuzzy rules in an answer group). I think it's definitely worth testing all possible situations of this (zero fuzzy rules, one fuzzy rule, two fuzzy rules) and ensuring the proper behavior happens.

Moreover, the code here is creating a policy that multiple fuzzy rules may be evaluated and contribute to the classification of this answer group, but may not participate in the best-match selection process. If you notice, all other situations where you're checking for a fuzzy rule you don't do a short-circuit to skip any other fuzzy rules in that answer group. I think you should be consistent with whether you are checking all fuzzy rules or just the first one. If you want to add the training data for all fuzzy rules but assume the first (or last) fuzzy rule is the best match, it should be tested to guard against changed behavior in the future.

Another approach that might actually be better is adding a validation check in exp_domain.AnswerGroup.validate to ensure explorations only have zero or one fuzzy rules per answer group, then have a corresponding test in exp_domain_test to test it. I actually lean towards this one since it better defines the behavior in this function and allows us to safely make the assumption there's exactly one fuzzy rule everywhere else in the code base as well.